### PR TITLE
[Rust] schema coerce and vector column inference

### PR DESCRIPTION
**Pull request by eddyxu**
Split the rust core from #466 for easy review and less merge conflicts.

### CODE REVIEW

File: rust/vectordb/src/data/sanitize.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    batch: RecordBatch,
+    schema: Arc<Schema>,
+) -> std::result::Result<RecordBatch, ArrowError> {
+    if batch.schema() == schema {
+        return Ok(batch);
+    }
+    let columns = schema
+        .fields()
+        .iter()
+        .flat_map(|field| {
```
Comment (wjones127) [inline rust/vectordb/src/data/sanitize.rs:141]: What is this `flat_map` and not `map`?

File: rust/vectordb/src/data/sanitize.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+) -> Result<Box<dyn RecordBatchReader + Send>> {
+    if reader.schema() == schema {
+        return Ok(Box::new(RecordBatchIterator::new(
+            reader.into_iter().collect::<Vec<_>>(),
+            schema,
+        )));
+    }
+    let batches = reader
+        .map(|batch| coerce_schema_batch(batch?, schema.clone()))
+        .collect::<Vec<_>>();
```
Comment (wjones127) [inline rust/vectordb/src/data/sanitize.rs:167]: Why collect? Can we just keep as an iterator?
```suggestion
    let batches = reader
        .map(|batch| coerce_schema_batch(batch?, schema.clone()));
```

File: rust/vectordb/src/data/sanitize.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+///
+pub fn coerce_schema(
+    reader: impl RecordBatchReader,
+    schema: Arc<Schema>,
+) -> Result<Box<dyn RecordBatchReader + Send>> {
+    if reader.schema() == schema {
+        return Ok(Box::new(RecordBatchIterator::new(
+            reader.into_iter().collect::<Vec<_>>(),
+            schema,
+        )));
```
Comment (wjones127) [inline rust/vectordb/src/data/sanitize.rs:163]: Similarly here, why collect and wrap it again:
```suggestion
        return Ok(Box::new(reader));
```

File: rust/vectordb/src/data/sanitize.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+                let actual_sub = array.as_fixed_size_list();
+                let values = coerce_array(actual_sub.values(), exp_field)?;
+                Ok(Arc::new(FixedSizeListArray::try_new_from_values(
+                    values.clone(),
+                    *dim,
+                )?) as Arc<dyn Array>)
+            }
+            DataType::List(_) => {
+                let list_arr = array.as_list::<i32>();
+                if list_arr.values().len() != (*exp_dim as usize) * list_arr.len() {
```
Comment (wjones127) [inline rust/vectordb/src/data/sanitize.rs:101]: hmm this isn't strictly sufficient. To be fully robust you'd need to check the length of each list. But maybe this is fine for now, given that if we reach this branch the user has already said they intend for a FSL of a certain length, right?

**Review by wjones127 — CHANGES_REQUESTED** [review]
I think you can avoid collecting in the reader.

File: rust/vectordb/src/data/sanitize.rs (hunk 1 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs:141]: `cargo clippy` suggested this way.

File: rust/vectordb/src/data/inspect.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+use arrow_schema::DataType;
+
+use crate::error::{Error, Result};
+
+/// Infer the vector columns from a dataset.
+///
+/// Parameters
+/// ----------
+/// - reader: RecordBatchReader
+/// - strict: if set true, only fixed_size_list<float> is considered as vector column. If set to false,
```
Comment (changhiskhan) [inline rust/vectordb/src/data/inspect.rs:46]: I'm wondering if we should have a minimum length requirement here. I guess UMAP/PCA results can be low dimensions, but I also worry about false positives?

File: rust/vectordb/src/data/inspect.rs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+pub fn infer_vector_columns(
+    reader: impl RecordBatchReader + Send,
+    strict: bool,
+) -> Result<Vec<String>> {
+    let mut columns = vec![];
+
+    let mut columns_map: HashMap<String, Option<u32>> = HashMap::new();
+    for field in reader.schema().fields() {
+        match field.data_type() {
+            DataType::FixedSizeList(sub_field, _) if sub_field.data_type().is_floating() => {
```
Comment (changhiskhan) [inline rust/vectordb/src/data/inspect.rs:57]: i forgot does `is_floating` check only for f32 or any float?

File: rust/vectordb/src/data/sanitize.rs (hunk 2 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs:167]: Fixed. Need to fix lifetime as well.

File: rust/vectordb/src/data/sanitize.rs (hunk 3 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs:163]: Fixed.

File: rust/vectordb/src/data/sanitize.rs (hunk 4 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs:101]: Oh, right. This was a directly copy from python sanitizer.  Lemme fix it.

File: rust/vectordb/src/data/inspect.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                columns.push(field.name().to_string());
+            }
+            DataType::List(sub_field) if sub_field.data_type().is_floating() && !strict => {
```
Comment (changhiskhan) [inline rust/vectordb/src/data/inspect.rs:60]: what about LargeListArray? safe to ignore for now?

File: rust/vectordb/src/data/sanitize.rs (hunk 4 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs]: Fixed.

File: rust/vectordb/src/data/inspect.rs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+    for batch in reader {
+        let batch = batch?;
+        let col_names = columns_map.keys().cloned().collect::<Vec<_>>();
+        for col_name in col_names {
+            let col = batch.column_by_name(&col_name).ok_or(Error::Schema {
+                message: format!("Column {} not found", col_name),
+            })?;
+            if let Some(dim) = infer_dimension(col.as_ref())? {
```
Comment (changhiskhan) [inline rust/vectordb/src/data/inspect.rs:73]: this is not in the query path right? Feels like the performance penalty might be big here?

File: rust/vectordb/src/data/inspect.rs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        return Ok(None);
+    };
+    let list_arr = arr.as_list_opt::<i32>().expect("must be a list now");
+    let len_arr = length(list_arr)?;
+    if len_arr.is_empty() {
+        return Ok(Some(0));
+    }
+
+    let dim = len_arr.as_primitive::<Int32Type>().value(0);
+    if bool_and(&eq_dyn_scalar(len_arr.as_primitive::<Int32Type>(), dim)?) != Some(true) {
```
Comment (changhiskhan) [inline rust/vectordb/src/data/inspect.rs:35]: i assume this short circuits right?

File: rust/vectordb/src/data/inspect.rs (hunk 6 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:57]: `is_floating` is `f16 | f32 | f64`

https://github.com/apache/arrow-rs/blob/65edbb1702e25420aacebe656dcd789690f72c82/arrow-schema/src/datatype.rs#L361

We need later to extend it to bf16 tho.

File: rust/vectordb/src/data/inspect.rs (hunk 7 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:60]: I can add it in this PR for safety.

File: rust/vectordb/src/data/inspect.rs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+use std::collections::HashMap;
+
+use arrow::compute::kernels::{aggregate::bool_and, length::length};
+use arrow_array::{cast::AsArray, types::Int32Type, Array, RecordBatchReader};
+use arrow_ord::comparison::eq_dyn_scalar;
+use arrow_schema::DataType;
+
+use crate::error::{Error, Result};
+
+pub(crate) fn infer_dimension(arr: &dyn Array) -> Result<Option<i32>> {
```
Comment (chebbyChefNEQ) [inline rust/vectordb/src/data/inspect.rs:24]: should we make this a generic function so we can customize the data width?

File: rust/vectordb/src/data/inspect.rs (hunk 8 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:73]: No, this is on the ingestion path. This is not slower than what we have in python `table::create()` now.

File: rust/vectordb/src/data/inspect.rs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    if !matches!(arr.data_type(), DataType::List(_)) {
+        return Ok(None);
+    };
+    let list_arr = arr.as_list_opt::<i32>().expect("must be a list now");
```
Comment (chebbyChefNEQ) [inline rust/vectordb/src/data/inspect.rs:28]: nit: iiuc, this `i32` really just means 32-bit wide, doesn't mean the data is actually integer. Could we alias this to be more clear?

```
type BIT_WIDTH_16 = i16;
...
```

File: rust/vectordb/src/data/inspect.rs (hunk 5 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:46]: 32? 64? I dont have a good sense where to cut the line now.

File: rust/vectordb/src/data/inspect.rs (hunk 11 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:28]: i32 means it is i32 as offset (which is `List` not `LargeList`).

File: rust/vectordb/src/data/inspect.rs (hunk 10 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:24]: I can just make it work with both list/large list? the data width of each element should be not relevant i think.

File: rust/vectordb/src/data/inspect.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+/// - reader: RecordBatchReader
+/// - strict: if set true, only fixed_size_list<float> is considered as vector column. If set to false,
+///           a list<float> column with same length is also considered as vector column.
+pub fn infer_vector_columns(
+    reader: impl RecordBatchReader + Send,
+    strict: bool,
+) -> Result<Vec<String>> {
+    let mut columns = vec![];
+
+    let mut columns_map: HashMap<String, Option<i32>> = HashMap::new();
```
Comment (chebbyChefNEQ) [inline rust/vectordb/src/data/inspect.rs:54]: nit: should we call this `columns_to_infer`?

File: rust/vectordb/src/data/inspect.rs (hunk 10 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:24]: https://docs.rs/arrow-array/46.0.0/arrow_array/array/struct.GenericListArray.html

File: rust/vectordb/src/data/sanitize.rs [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        .fields()
+        .iter()
+        .flat_map(|field| {
+            batch
+                .column_by_name(field.name())
+                .map(|c| coerce_array(c, field))
+                .ok_or(|| {
+                    ArrowError::SchemaError(format!("Column {} not found in batch", field.name()))
+                })
+        })
```
Comment (wjones127) [inline rust/vectordb/src/data/sanitize.rs:158]: Try this instead:
```suggestion
        .map(|field| {
            batch
                .column_by_name(field.name())
                .ok_or(|| {
                    ArrowError::SchemaError(format!("Column {} not found in batch", field.name()))
                })
                .and_then(|c| coerce_array(c, field))
        })
```

File: rust/vectordb/src/data/sanitize.rs [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+) -> std::result::Result<Arc<dyn Array>, ArrowError> {
+    if array.data_type() == field.data_type() {
+        return Ok(array.clone());
+    }
+    match (array.data_type(), field.data_type()) {
+        // Normal cast-able types.
+        (adt, dt) if can_cast_types(adt, dt) => cast(&array, dt),
+        // Casting between f16/f32/f64 can be lossy.
+        (adt, dt) if (adt.is_floating() || dt.is_floating()) => {
+            if adt.byte_width() < dt.byte_width() {
```
Comment (chebbyChefNEQ) [inline rust/vectordb/src/data/sanitize.rs:75]: should this be flipped?

array (32-bit) < field (64-bit) should be lossless

File: rust/vectordb/src/data/sanitize.rs (hunk 1 above)
Comment (wjones127) [reply rust/vectordb/src/data/sanitize.rs:151]: Oh I see it's a nested result. I don't think you should call `flatten()` on result; it ignores the errors.

File: rust/vectordb/src/data/inspect.rs (hunk 11 above)
Comment (chebbyChefNEQ) [reply rust/vectordb/src/data/inspect.rs:28]: Ah ok that makes sense

File: rust/vectordb/src/data/sanitize.rs (hunk 13 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs:158]: This results to a `Result<Vec<Result<_, _>>` ?

File: rust/vectordb/src/data/inspect.rs (hunk 12 above)
Comment (eddyxu) [reply rust/vectordb/src/data/inspect.rs:54]: Done.

File: rust/vectordb/src/data/sanitize.rs (hunk 14 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs:75]: Good catch.

File: rust/vectordb/src/data/sanitize.rs (hunk 13 above)
Comment (eddyxu) [reply rust/vectordb/src/data/sanitize.rs]: ok, made it work.  Change to use `ok_or(ArrowError::...)` directly

File: rust/vectordb/src/data/sanitize.rs [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+    let columns = schema
+        .fields()
+        .iter()
+        .map(|field| {
+            batch
+                .column_by_name(field.name())
+                .map(|c| coerce_array(c, field))
+                .ok_or(ArrowError::SchemaError("Column not found".to_string()))
+        })
+        .collect::<std::result::Result<std::result::Result<Vec<_>, ArrowError>, ArrowError>>()??;
```
Comment (wjones127) [inline rust/vectordb/src/data/sanitize.rs:170]: Sorry, I didn't mean for you to nest the `Result`. This time I've tested this locally:
```suggestion
        .map(|field| {
            batch
                .column_by_name(field.name())
                .ok_or_else(|| {
                    ArrowError::SchemaError(format!("Column {} not found", field.name()))
                })
                .and_then(|c| coerce_array(c, field))
        })
        .collect::<std::result::Result<Vec<_>, ArrowError>>()?;
```

**Review by wjones127 — APPROVED** [review]

**Pull request merged**